mono/mini/mini-windows-tls-callback.c

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright 2019 Microsoft Corporation
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+ */
+
+#include "mini-runtime.h"
+
+#if defined(HOST_WIN32)
+#include "mini-windows.h"
+#include <windows.h>
+
+/* Only reset by initial callback from OS loader with reason DLL_PROCESS_ATTACH */
+/* both for DLL and LIB callbacks, so no need to protect this variable. */
+static MonoWin32TLSCallbackType mono_win32_tls_callback_type = MONO_WIN32_TLS_CALLBACK_TYPE_NONE;
+
+/* NOTE, this function needs to be in this source file to make sure linker */
+/* resolve this symbol from mini-windows.c, picking up callback and */
+/* included it in TLS Directory PE header. Function makes sure we only activate */
+/* one of the supported mono callback types at runtime (if multiple one has been used) */
+gboolean
+mono_win32_handle_tls_callback_type (MonoWin32TLSCallbackType callback_type)
+{
+	if (mono_win32_tls_callback_type == MONO_WIN32_TLS_CALLBACK_TYPE_NONE)
+		mono_win32_tls_callback_type = callback_type;
+
+	if (callback_type != mono_win32_tls_callback_type)
+		return FALSE;
+
+	return TRUE;
+}
+
+MONO_EXTERN_C
+VOID NTAPI mono_win32_tls_callback (PVOID module_handle, DWORD reason, PVOID reserved);
+
+MONO_EXTERN_C
+VOID NTAPI mono_win32_tls_callback (PVOID module_handle, DWORD reason, PVOID reserved)
+{
+	mono_win32_runtime_tls_callback ((HMODULE)module_handle, reason, reserved, MONO_WIN32_TLS_CALLBACK_TYPE_LIB);
+}
+
+/* If we are building a static library that won't have access to DllMain we can't */
+/* correctly detach a thread before it terminates. The MSVC linker + runtime (also applies to MINGW) */
+/* uses a set of predefined segments/sections where callbacks can be stored and called */
+/* by OS loader (part of TLS Directory PE header), regardless if runtime is used */
+/* as a static or dynamic library. */
+#if (_MSC_VER >= 1400)
+#pragma const_seg(".CRT$XLX")
+MONO_EXTERN_C const PIMAGE_TLS_CALLBACK __mono_win32_tls_callback = mono_win32_tls_callback;
+#pragma const_seg()
```

Comment:
There is also `__declspec(allocate)` but I'm not sure the advantages/disadvantages.

Reply:
Does similar job, but a little harder to open/close sections using that method (depending on where you are in the source), but the result is similar.